File: 314-binary-tree-vertical-order-traversal/binary-tree-vertical-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        # Dictionary to hold nodes in each column
        column_table = defaultdict(list)
        # Queue for BFS; stores tuples of (node, column_index)
        queue = deque([(root, 0)])
        
        # Perform BFS traversal
        while queue:
            node, col = queue.popleft()
            if node:
                # Add the node's value to the list of its respective column index
                column_table[col].append(node.val)
                # Queue the left and right children with updated column indices
                queue.append((node.left, col - 1))
                queue.append((node.right, col + 1))
        
        # Sort the columns and retrieve the nodes in column order
        return [column_table[x] for x in sorted(column_table.keys())]

File: 314-binary-tree-vertical-order-traversal/binary-tree-vertical-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        // map maps column to the nodes in that column
        Map<Integer, List<Integer>> map = new TreeMap<>(); 
        // pair<node, col> 
        Queue<Pair<TreeNode, Integer>> q = new LinkedList<>(); 
        if (root != null) q.offer(new Pair<>(root, 0)); 
        while (!q.isEmpty()) {
            Pair<TreeNode, Integer> curr = q.poll(); 
            TreeNode node = curr.getKey(); 
            int col = curr.getValue(); 
            map.putIfAbsent(col, new ArrayList<>()); 
            map.get(col).add(node.val);
            if (node.left != null) q.offer(new Pair<>(node.left, col - 1)); 
            if (node.right != null) q.offer(new Pair<>(node.right, col + 1)); 
        }
        return new ArrayList<>(map.values()); 
    }
}